core/templates/pages/exploration-editor-page/statistics-tab/charts/bar-chart.component.ts

```diff
@@ -68,6 +70,15 @@ angular.module('oppia').directive('barChart', [function() {
 
       $scope.$watch('data()', redrawChart);
       $(window).resize(redrawChart);
-    }]
-  };
-}]);
+    };
+
+    // The destroy lifecycle hook is being used here to remove resize event
+    // listener because google charts api is only accessible in exploration
+    // editor page. Without this hook, the resize callback may be triggered
+    // by other components that is not in exploration editor page, raising
+    //  an error. This is both for the application itself and unit tests.
+    ctrl.$onDestroy = function() {
+      $(window).off('resize');
```

Comment:
One question about this. Suppose other window.resize event listeners have been added by other components. Will this destroy them all too?

(In your reply please explain the steps you took, when investigating, to come to a conclusion.)

Reply:
I don't think so. I've tested it in exploration-footer and it's working as expected. Also, I don't think this can cause a problem because components are created multiple times when changing pages.

Reply:
Can you explain the details of your test (as requested in the original comment) so that I can try to reproduce it?

Reply:
Here's what I did:
- I've put a console.log in exploration-footer.directive on line 51.
- I've run the application and created an exploration with two states.
- I've visited the exploration with 2 different users.
- After, I came back to the exploration editor page in preview tab. Then, I resized the page and the console.log appeared.
- I changed to statistics tab and the pie chart was there. Then, I changed back to preview tab and I did the resize again, the console.log appeared. So this means that it's still working after pie chart was destroyed.

The window dimensions service (the one that triggers resize event), was one of the services that were breaking the unit tests before these changes.

Reply:
That's a good test in terms of structure, but I'm a bit worried about the details because the way resizing is handled in exploration-footer is not the same as the way it is handled in bar-chart.component.ts. Can you try this again, but add an additional resize handler in exploration-footer that is similar to the bar-chart one?

If that ends up working, great. If not, then maybe converting the bar-chart one to follow the exploration-footer one might work, perhaps.

Reply:
Just to add, the `top-navigation-bar` and `editor-navigation` directives use `WindowDimensionsService`'s resize listener. Not sure if this would affect those, but might be  worth checking, or maybe use the functions in that service for checking for 'resize' here?

Reply:
I've checked this situation but I'm confused. `ctrl.$onDestroy` is not being called when the chart is destroyed. According to this [comment](https://github.com/angular/angular.js/issues/15073#issuecomment-245270829), it's because the component's `$scope` is not from the controller itself, but from its parent. So I think this hook is only triggered when `exploration-editor-page` is destroyed. However, I see that when changing routes, we reload the page, so we don't need to handle destroy hooks.

`ctrl.$onDestroy` is being used only for unit test purposes. I think it's not the perfect approach, but it works. I don't know how to solve this specific situation without using a destroy hook.

Reply:
Based on the link you provided, shouldn't you be using $scope.$on('destroy') here instead?

Reply:
Also, did you do any research on how to remove specific listeners from the resize hook, rather than clearing all listeners? I think that's really the main open question here.

Reply:
`$scope.$on('$destroy')` didn't work as well. After doing some checks and local tests, I see that the resize event doesn't affect other resize events, otherwise we would have realized a problem before. I've searched how to remove specific listeners and I didn't find any concrete approach. However, I found a solution that it's better than the actual one, PTAL!

Reply:
Hi @marianazangrossi, I don't understand, sorry. What do you mean by "I see that the resize event doesn't affect other resize events"? If you explain more clearly the tests that you did to confirm this, it would be easier to understand what behaviour you're observing.

Reply:
Also -- I just realized that @aks681 gave a solution and I'm not sure if you investigated it. Could you take a look at the resize mechanisms used by the other directives that he mentioned? Normally we don't use jQuery $(window) and that might be what's causing all the issues here.

I think you can subscribe() and unsubscribe() to events. See the Angular documentation for observables. Perhaps have a quick chat with @srijanreddy98 if you need help with this.

Reply:
Hi @seanlip, I think I wasn't so clear. What I meant is that a specific resize callback is not affected by other resizes callback. For example, the callback for `editor-navigator` does not affect the callback for `pie-chart` and vice-versa. Also, I think these resize callbacks from `exploration-editor-page` are destroyed only when changing to another page.

However, I've changed it to subscribes and I think it's much better, thanks for the suggestion! PTAL.